### ISSUE TITLE
Freezing a certain layer during the training process

### ISSUE THREAD

**Issue by cainiao123s** [ultralytics/ultralytics#20136]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, may I ask if during the training process, I can set to freeze a certain layer only when epochs=10? The layer will not be frozen in other epochs.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cainiao123s, thank you for your interest in Ultralytics 🚀! We recommend exploring the [Docs](https://docs.ultralytics.com/) for guidance where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many of the most common questions may already be answered there.

If this is a custom training ❓ Question, please provide as much information as possible to assist us in understanding your use case, such as a description of the model, training logs, and dataset specifics. You can also check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for optimization advice.

If this is a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to allow us to debug the issue effectively.

Join the Ultralytics community where it suits you best:
- Real-time chat? Join us on [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions? Check out [Discourse](https://community.ultralytics.com/)
- Engaging community threads? Visit our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Ensure you are using the latest `ultralytics` package and verify your issue is not already resolved. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these up-to-date verified environments with all dependencies preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests validate the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

This is an automated response 🛠️, but an Ultralytics engineer will assist you shortly! 😊

**Comment by Y-T-G** [comment]
You can do that using callbacks.

https://github.com/ultralytics/ultralytics/blob/e19f0ef606c67f028a998b23223ead689ad4c8c6/ultralytics/engine/trainer.py#L431

**Comment by NOTGOOOOD** [comment]
> You can do that using callbacks.
> 
> [ultralytics/ultralytics/engine/trainer.py](https://github.com/ultralytics/ultralytics/blob/e19f0ef606c67f028a998b23223ead689ad4c8c6/ultralytics/engine/trainer.py#L431)
> 
> Line 431 in [e19f0ef](/ultralytics/ultralytics/commit/e19f0ef606c67f028a998b23223ead689ad4c8c6)
> 
>  self.run_callbacks("on_train_epoch_end")

Sorry, I don't understand also. How can I implement the callback function? And where should I use the callback function? My training code as below:

```Python
model = YOLO()
model.train(data="mydata.yaml")
```

**Comment by glenn-jocher** [comment]
You can use callbacks to freeze/unfreeze layers at specific epochs. Here's how to implement it with your training code:

```python
def conditional_freeze(trainer):
    current_epoch = trainer.start_epoch + len(trainer.tloss) if hasattr(trainer, 'tloss') else 0
    
    if current_epoch == 10:
        # Freeze specific layers (example: freeze first 5 layers)
        for i, (name, param) in enumerate(trainer.model.named_parameters()):
            if i < 5:  # adjust layer selection as needed
                param.requires_grad = False
    else:
        # Unfreeze all layers
        for param in trainer.model.parameters():
            param.requires_grad = True

model = YOLO()
model.add_callback("on_train_epoch_start", conditional_freeze)
model.train(data="mydata.yaml")
```

The callback function receives a `trainer` object and you can use `model.add_callback()` to register it before training. You can find all available callback events in the [callbacks documentation](https://docs.ultralytics.com/usage/callbacks/).

**Comment by NOTGOOOOD** [comment]
@glenn-jocher  Thank you for your help ！

**Comment by glenn-jocher** [comment]
You're welcome! Happy to help with the callback implementation. Feel free to ask if you have any other questions about training customization.
